osu.Framework/Allocation/DependencyContainer.cs

```diff
@@ -17,142 +14,105 @@ public class DependencyContainer : IReadOnlyDependencyContainer
 
         /// <summary>
         /// Create a new DependencyContainer instance.
         /// </summary>
         /// <param name="parent">An optional parent container which we should use as a fallback for cache lookups.</param>
-        public DependencyContainer(IReadOnlyDependencyContainer parent = null)
+        public DependencyContainer(IReadOnlyDependencyContainer? parent = null)
         {
             parentContainer = parent;
         }
 
         /// <summary>
-        /// Caches an instance of a type as its most derived type. This instance will be returned each time you <see cref="Get(Type)"/>.
+        /// Caches an instance of a type as its most derived type. This instance will be returned each time you <see cref="Get{T}()"/>.
         /// </summary>
         /// <param name="instance">The instance to cache.</param>
         public void Cache(object instance)
             => Cache(instance, default);
 
         /// <summary>
-        /// Caches an instance of a type as its most derived type. This instance will be returned each time you <see cref="Get(Type)"/>.
+        /// Caches an instance of a type as its most derived type. This instance will be returned each time you <see cref="Get{T}()"/>.
         /// </summary>
         /// <param name="instance">The instance to cache.</param>
         /// <param name="info">Extra information to identify <paramref name="instance"/> in the cache.</param>
         public void Cache(object instance, CacheInfo info)
-            => CacheAs(instance.GetType(), info, instance, false);
-
-        /// <summary>
-        /// Caches an instance of a type as a type of <typeparamref name="T"/>. This instance will be returned each time you <see cref="Get(Type)"/>.
-        /// </summary>
-        /// <param name="instance">The instance to cache. Must be or derive from <typeparamref name="T"/>.</param>
-        public void CacheAs<T>(T instance) where T : class
-            => CacheAs(instance, default);
+        {
+            ArgumentNullException.ThrowIfNull(instance);
 
-        /// <summary>
-        /// Caches an instance of a type as a type of <typeparamref name="T"/>. This instance will be returned each time you <see cref="Get(Type)"/>.
-        /// </summary>
-        /// <param name="instance">The instance to cache. Must be or derive from <typeparamref name="T"/>.</param>
-        /// <param name="info">Extra information to identify <paramref name="instance"/> in the cache.</param>
-        public void CacheAs<T>(T instance, CacheInfo info) where T : class
-            => CacheAs(typeof(T), info, instance, false);
+            CacheAs(instance.GetType(), info, instance);
+        }
 
         /// <summary>
-        /// Caches an instance of a type as a type of <paramref name="type"/>. This instance will be returned each time you <see cref="Get(Type)"/>.
+        /// Caches an instance of a type as its most derived type. This instance will be returned each time you <see cref="Get{T}()"/>.
```

Comment:
If we're editing these, I'd probably clarify further to say

```suggestion
        /// Caches an instance of a type as its most derived type as returned by <see cref="object.GetType"/>. This instance will be returned each time you <see cref="Get{T}()"/>.
```

(goes for all other overloads of this)